### ISSUE TITLE
Cross Attention Query

### ISSUE THREAD

**Issue by lhallee** [agemagician/ProtTrans#95]
Hi! When I use ProtBert as a tokenizer, I get a BaseModelOutputWithPoolingAndCrossAttentions with some tensors and cross_attentions = None. What do these cross_attentions refer to? and how do I get their output?

**Comment by mheinzinger** [comment]
Hi, how do you load the ProtBERT-Tokenizer? If you have a seq2seq Transformer with Encoder and Decoder, cross-attention connects the output of the Encoder to the input of the Decoder to condition the generated sequence on your input. BERT is an encoder-only model so it does not have any cross-attention which is why it makes sense that this value is None.

**Comment by lhallee** [comment]
Hi! That's very helpful, thank you. I think I will switch to ProtT5. I'm trying to embed amino acid sequences with the following tutorial from you guys: https://github.com/agemagician/ProtTrans/blob/master/Embedding/PyTorch/Advanced/ProtT5-XL-BFD.ipynb . It works great with small datasets, but I rapidly run out of VRAM when scaling up. Is there a way to batch this process? 100 sequences appear to require upwards of 50GB of VRAM. Considering I'd like to perform this task on 1000s it seems I'll need many small batches.

**Comment by mheinzinger** [comment]
hm, yes, you have to watch out for multiple things when creating batches as longer sequences take quadratically more memory so you can not have a fixed number of sequences per batch but rather some amino acid count.
Our [colab](https://colab.research.google.com/drive/1TUj-ayG3WO52n5N50S7KH9vtt6zRkdmj?usp=sharing#scrollTo=nK4hwGggR_Rs) has an example on how to batch-process with ProtT5.

**Comment by Moeinh77** [comment]
Hi, I was wondering if cross-attention can show any specific relationship between the two input sequences. Also, If I use encode plus and encode two sequences, in your experience will it show any meaningful relationships between two protein sequences based on the generated attention matrix which is in the dimensions of [len(seq1)+len(seq2)] x [len(seq1)+len(seq2)]? I'd appreciate it if you guide me.

**Comment by mheinzinger** [comment]
Hi, very interesting idea! - In fact, we did not try yet how our models behave when we just concatenate two sequences and feed them in. Just some thoughts:
- If you concatenate two sequences and feed them both to the encoder of ProtT5, you rather get a a mixed attention that sees both sequences. If I think about cross-attention in this context, I rather imagine that you feed in one sequences to the encoder, one to the decoder (similar to a translation task) and observe how the cross-attention mixes the two. So the model first attends to the source-language (Protein P1 with L1) and then uses cross-attention to mix it with the target-language sentence (Protein P2 with L2). However, I would be surprised if this worked because we trained solely on denoising, not on translation.

- What you can always do (irrespective of cross-attention): simply concatenate the two sequences and feed them into the encoder. By checking the section of the attention matrix that attends from L1 to L2 (or vice versa), you might get a signal on their relationship/interaction_probability.

**Comment by Moeinh77** [comment]
Thank you for sharing your thoughts. I wonder since the ProtTrans models are not trained for next-sentence prediction, does seq-to-seq attention can include any sensible info solely with Masked Language Modeling pretraining.

**Comment by mheinzinger** [comment]
Hm, this is an open Q. Intuitively I would lean towards: No, there is probably no information.
On the other hand: why not give it a shot an confirm it. After all, AlphaFold2 also learnt to predict PPIs to a certain extent. I absolutely do not want to claim that our work is in any way comparable to AlphaFold2 but the underlying idea that intra-protein contacts might share some similarity to inter-protein contacts/interactions might still work, at least, to a certain extent.

So no: I doubt that you will find much there but if you only found something, that would already be interesting :)

**Comment by lhallee** [comment]
How does training work in general for the t5 model? The typical Seq2Seq output does not include loss, so how would one fine-tune your ProtT5 with MLM or NSP given this output?

**Comment by mheinzinger** [comment]
Hm, I am not sure whether I misunderstand your Q. There is a loss, otherwise, there would not be a backpropagation signal for seq2seq models.  The usual cross-entropy usually does this job for you in case of seq2seq models. The difficult bit is just to handle spans of tokens with variable length in case of T5.

When it comes to fine-tuning, you can simply start from the pre-trained ProtT5 model as a starting point. From there on, you can re-design the objective and apply the loss that fits your needs. For example, in NLP people first pre-train on unlabeled data and afterwards finetune on a small subset of high-quality translations. In this case, you would have very different objectives, i.e., denoising during pre-training and translation during fine-tuning. Similarly, you can see ProtT5 as a better way to initialize your new model (better compared to random initialiization).
